File: data-template/wordcount.py
"""Pure word/character count for a rendered reading fragment (content/<id>.html).

Counts the AUTHOR'S PROSE: headings, body text, figure/table captions. Excludes the References /
bibliography block, the footnotes list, and math (equations are not words). Regex-based on purpose —
the input is a well-formed pandoc HTML5 fragment and this must run without extra dependencies.
"""
import re

# Blocks excluded WITH their nested content: the bibliography and the footnotes list. citeproc emits
# <div id="refs" class="references csl-bib-body"> wrapping NESTED <div class="csl-entry"> — a non-greedy
# regex stops at the FIRST inner </div> and leaks every following reference into the count, so match the
# closing tag by BALANCED depth instead. Attribute values may be single- or double-quoted.
_REF_OPEN = re.compile(r'<(div|section)\b[^>]*\bid=["\']refs["\'][^>]*>', re.I)
_REFCLASS_OPEN = re.compile(r'<(div|section)\b[^>]*\bclass=["\'][^"\']*\breferences\b[^"\']*["\'][^>]*>', re.I)
_FN_OPEN = re.compile(r'<(section|div|aside)\b[^>]*\bclass=["\'][^"\']*\bfootnotes\b[^"\']*["\'][^>]*>', re.I)
# math stays as <span class="math ...">\(...\)</span> in server HTML (KaTeX renders client-side); leaf, no nesting.
_MATH_RE = re.compile(r'<span\b[^>]*\bclass=["\'][^"\']*\bmath\b[^"\']*["\'][^>]*>.*?</span>', re.I | re.S)
_TAG_RE = re.compile(r'<[^>]+>')
_ENT_RE = re.compile(r'&[a-zA-Z]+;|&#\d+;')


def _strip_balanced(s, opener):
    """Remove every element matched by `opener` INCLUDING nested content, finding the matching close by
    counting same-name open/close tags (regex alone can't balance nested divs like a citeproc bibliography)."""
    while True:
        m = opener.search(s)
        if not m:
            return s
        tag = m.group(1)
        close = re.compile(r'<(/?)' + tag + r'\b[^>]*>', re.I)
        depth, end = 1, None
        for t in close.finditer(s, m.end()):
            depth += -1 if t.group(1) else 1
            if depth == 0:
                end = t.end()
                break
        s = s[:m.start()] + " " + (s[end:] if end is not None else "")


def word_count(html):
    s = html or ""
    for opener in (_REF_OPEN, _REFCLASS_OPEN, _FN_OPEN):
        s = _strip_balanced(s, opener)
    s = _MATH_RE.sub(" ", s)
    s = _TAG_RE.sub(" ", s)
    s = _ENT_RE.sub(" ", s)
    words = s.split()
    return {"words": len(words), "chars": len(" ".join(words))}   # chars WITH spaces (Word's default metric)

File: js/wordcount.js
// Word-count display helpers + a client-side fallback counter. The fallback mirrors data-template/
// wordcount.py so a project without counts.json (not re-rendered yet) still shows a count. Term-neutral.

export function formatCount(n) {
  const v = Number(n) || 0;
  return `${v.toLocaleString('en-US')} word${v === 1 ? '' : 's'}`;
}

export function totalWords(counts = {}) {
  return Object.values(counts || {}).reduce((s, c) => s + (c && c.words ? c.words : 0), 0);
}
export function totalChars(counts = {}) {
  return Object.values(counts || {}).reduce((s, c) => s + (c && c.chars ? c.chars : 0), 0);
}

// Fallback: same rules as the engine (refs / footnotes / math excluded), quote-tolerant. Reference and
// footnote blocks are stripped by BALANCED tag depth (citeproc's <div id="refs"> nests same-name divs, so a
// non-greedy regex would stop at the first inner </div> and leak the rest). Openers keep group 1 = tag name.
const REF_OPEN = '<(div|section)\\b[^>]*\\bid=["\']refs["\'][^>]*>';
const REFCLASS_OPEN = '<(div|section)\\b[^>]*\\bclass=["\'][^"\']*\\breferences\\b[^"\']*["\'][^>]*>';
const FN_OPEN = '<(section|div|aside)\\b[^>]*\\bclass=["\'][^"\']*\\bfootnotes\\b[^"\']*["\'][^>]*>';
const MATH = /<span\b[^>]*\bclass=["'][^"']*\bmath\b[^"']*["'][^>]*>[\s\S]*?<\/span>/gi;   // leaf, no nesting
const TAG = /<[^>]+>/g;
const ENT = /&[a-zA-Z]+;|&#\d+;/g;

function stripBalanced(s, openerSrc) {
  for (;;) {
    const m = new RegExp(openerSrc, 'i').exec(s);
    if (!m) return s;
    const close = new RegExp('<(/?)' + m[1] + '\\b[^>]*>', 'gi');
    close.lastIndex = m.index + m[0].length;
    let depth = 1, end = -1, t;
    while ((t = close.exec(s))) { depth += t[1] ? -1 : 1; if (depth === 0) { end = close.lastIndex; break; } }
    s = s.slice(0, m.index) + ' ' + (end >= 0 ? s.slice(end) : '');
  }
}

export function countWords(html) {
  let s = String(html || '');
  s = stripBalanced(s, REF_OPEN);
  s = stripBalanced(s, REFCLASS_OPEN);
  s = stripBalanced(s, FN_OPEN);
  s = s.replace(MATH, ' ').replace(TAG, ' ').replace(ENT, ' ');
  const words = s.split(/\s+/).filter(Boolean);
  return { words: words.length, chars: [...words.join(' ')].length };   // chars WITH spaces (codepoint-accurate)
}

// A unit counts as "not counted yet" only when its words value isn't a number. A genuine 0 (an empty
// unit) is a real answer and must not be refetched forever — but it is also what an unopened unit shows
// before anything has counted it, which is why the count-all pass writes a number for every unit.
export function missingCountIds(units = [], counts = {}) {
  return (units || []).filter(u => u && u.id && typeof (counts?.[u.id]?.words) !== 'number').map(u => u.id);
}

// The engine's counts.json is authoritative (written at render time from the real built HTML); the local
// cache only fills units the engine hasn't published yet, so a stale cached number can never mask a fresh
// rendered one.
export function mergeCounts(engine, cached) {
  const ok = v => v && typeof v.words === 'number';
  const out = {};
  for (const [k, v] of Object.entries(cached || {})) if (ok(v)) out[k] = v;
  for (const [k, v] of Object.entries(engine || {})) if (ok(v)) out[k] = v;
  return out;
}
